### PR TITLE
Randomize KnnVector codec params in RandomCodec; addresses gh-14047

### PR DESCRIPTION
**Pull request by msokolov**
This adds the randomization to RandomCodec and addresses some test issues.

* In a few places I had to disable the randomization; basically wherever we are carefully testing scores. In these cases we won't have the same testing for quantized scores, but I think those would require more targeted testing since it's unclear what kinds of guarantees we can make about score values under quantization.
* In a few other palces I added some slop to the score comparisons; basically if we had a score comparison in a test that was primarily focused on other things (like testing Explain)
* I fixed a divide-by-zero edge case in the ScalarQuantizer (when there is only a single vector in a segment)
* I standardized bounds checking so we have consistent exception messages for dimension mismatches

### CODE REVIEW

**Comment by msokolov** [comment]
oops a test failed; my bad for only beasting the ones in core ... I'll look at this one:

    gradlew :lucene:join:test --tests "org.apache.lucene.search.join.TestBlockJoin.testSimpleKnn" -Ptests.jvms=2 -Ptests.jvmargs= -Ptests.seed=E773842CB2A422CA

**Comment by msokolov** [comment]
Another test fails now (this is going to be fun, it's hard to ferret all these things out proactively):

```
TestBasicBackwardsCompatibility > testIndexOldIndex {Lucene-Version:10.0.0; Pattern: index.%1$s-cfs.zip} FAILED
    org.junit.ComparisonFailure: wrong first document expected:<4[4]> but was:<4[1]>
        at __randomizedtesting.SeedInfo.seed([22D75A8559728A8D:93B6C125DE4C6E18]:0)
```

Looking at this test, I can't tell where these assertions came from: I think they might be too ambitious? @s1monw did you invent this?  The entire class was created out of whole cloth, but I assume at least some of these tests were pre-existing?

**Comment by s1monw** [comment]
@msokolov I didn't invent this test. Yet, we never tested the current version obeying with it's own BWC expecations. Please make sure nothing randomizes anything in this test. BWC tests are not made for this. you need to supress anything that potentially randomizes stuff here. 
> Another test fails now (this is going to be fun, it's hard to ferret all these things out proactively):

maybe revert then? That's what it takes I guess?

**Comment by msokolov** [comment]
Thank Simon, that's good feedback about avoiding randomization for these tests. There's nothing to revert - this hasn't been committed, it's just github running tests that is failing. I do wonder though if we run the risk of a user with a customized Codec (just applying settings like graph bushiness and enabling quantization requires a custom Codec, so it would be typical) failing to be able to load an older index version for some reason, and we'd have no advance warning from the back-compat tests.  Maybe that shouldn't be a concern, I'm not sure. In any case I will disable the randomization for these tests.

**Review by benwtrent — APPROVED** [review]
I do think this should have a "CHANGES" entry for the quantization bug fix. That bug will happen with uniform vector values which is common in testing infrastructure.

**Comment by msokolov** [comment]
makes sense, I'll add a CHANGES entry and commit

File: lucene/core/src/java/org/apache/lucene/util/quantization/ScalarQuantizer.java [hunk 1]
```diff
@@ -97,12 +97,18 @@ public ScalarQuantizer(float minQuantile, float maxQuantile, byte bits) {
     }
     assert maxQuantile >= minQuantile;
     assert bits > 0 && bits <= 8;
-    this.minQuantile = minQuantile;
-    this.maxQuantile = maxQuantile;
     this.bits = bits;
     final float divisor = (float) ((1 << bits) - 1);
-    this.scale = divisor / (maxQuantile - minQuantile);
-    this.alpha = (maxQuantile - minQuantile) / divisor;
+    if (minQuantile == maxQuantile) {
```
Comment (mccullocht) [inline lucene/core/src/java/org/apache/lucene/util/quantization/ScalarQuantizer.java:102]: @msokolov mind if I backport just this changes in this file into branch_10x tonight?

Comment (msokolov) [reply lucene/core/src/java/org/apache/lucene/util/quantization/ScalarQuantizer.java:102]: No, although it might be too late to do it on the "tonight" of which you were writing!
